File: structural/decorator/decorator.java
// Component
public interface Beverage {
    int getCost();
    String getTitle();
}

// ConcreteComponent
public class Americano implements Beverage {
    @Override
    public int getCost() {
        return 3000;
    }

    @Override
    public String getTitle() {
        return "Americano";
    }
}

// ConcreteComponent
public class HerbalTea implements Beverage {
    @Override
    public int getCost() {
        return 4000;
    }

    @Override
    public String getTitle() {
        return "Herb Tea";
    }
}

// Decorator
public abstract class CondimentDecorator implements Beverage {
    private final Beverage beverage;

    public CondimentDecorator(Beverage beverage) {
        this.beverage = beverage;
    }

    @Override
    public int getCost() {
        return beverage.getCost();
    }
    @Override
    public String getTitle() {
        return beverage.getTitle();
    }
}

// ConcreteDecorator
public class Sugar extends CondimentDecorator {
    public Sugar(Beverage beverage) {
        super(beverage);
    }

    @Override
    public int getCost() {
        return super.getCost() + 200;
    }

    @Override
    public String getTitle() {
        return super.getTitle() + " Add Sugar ";
    }
}

// ConcreteDecorator
public class Shot extends CondimentDecorator {
    public Shot(Beverage beverage) {
        super(beverage);
    }

    @Override
    public int getCost() {
        return super.getCost() + 500;
    }

    @Override
    public String getTitle() {
        return super.getTitle() + " Add Shot ";
    }
}

Beverage americano = new Americano();
Beverage americanoOneShot = new Shot(americano);

System.out.println(americanoOneShot.getCost());
System.out.println(americanoOneShot.getTitle());

Beverage herbalTea = new HerbalTea();
Beverage herbalTeaOneSugar = new Sugar(herbalTea);

System.out.println(herbalTeaOneSugar.getCost());
System.out.println(herbalTeaOneSugar.getTitle());

File: structural/decorator/decorator.py
from abc import ABCMeta, abstractmethod

class Beverage(metaclass=ABCMeta):
    """
    Component
    """
    @abstractmethod
    def get_cost(self) -> int:
        pass

    @abstractmethod
    def get_description(self) -> str:
        pass

class Americano(Beverage):
    """
    Concrete Component
    """
    def get_cost(self) -> int:
        return 3000

    def get_description(self) -> str:
        return 'Americano'

class HerbalTea(Beverage):
    """
    Concrete Component
    """
    def get_cost(self) -> int:
        return 4000

    def get_description(self) -> str:
        return 'Herb Tea'

class CondimentDecorator(Beverage):
    """
    Decorator
    """
    def __init__(self, beverage: Beverage):
        self.__beverage = beverage

    @abstractmethod
    def get_cost(self) -> int:
        return self.__beverage.get_cost()

    @abstractmethod
    def get_description(self) -> str:
        return self.__beverage.get_description()

class Sugar(CondimentDecorator):
    """
    Concrete Decorator
    """
    def __init__(self, beverage: Beverage):
        super().__init__(beverage)

    def get_cost(self) -> int:
        return super().get_cost() + 200

    def get_description(self) -> str:
        return super().get_description() + ' Add Sugar '

class Shot(CondimentDecorator):
    """
    Concrete Decorator
    """
    def __init__(self, beverage: Beverage):
        super().__init__(beverage)

    def get_cost(self) -> int:
        return super().get_cost() + 500

    def get_description(self) -> str:
        return super().get_description() + ' Add Shot '

americano = Americano()
americano_one_sugar = Sugar(americano)

print(americano_one_sugar.get_cost())
print(americano_one_sugar.get_description())

# 3200
# Americano + Sugar 
